File: apps/faithful-ios/Sources/FaithfulKit/Giving/GivingClient.swift
import Foundation

/// Reads the giving projections and starts gifts.
///
/// ## What is cached, and what is not
///
/// The **fund list** is cached under the caller's partition and revalidated with
/// its ETag, like every other projection: it is a church's published
/// configuration and it changes rarely.
///
/// **Nothing else is.** A donation session carries a client secret, a status is
/// the one thing where a stale answer is actively harmful, and a person's giving
/// history is not something to leave sitting in a cache on a shared device. All
/// three routes say `no-store`, and this client does not second-guess them.
public actor GivingClient {
    private let api: APIClient
    private let cache: PartitionedCache

    public init(api: APIClient, cache: PartitionedCache) {
        self.api = api
        self.cache = cache
    }

    // MARK: - Funds

    public func home(
        churchSlug: String,
        partition: CachePartition
    ) async throws -> GivingHome {
        let key = "giving.funds.\(churchSlug)"
        let cached = await cache.load(GivingHome.self, name: key, partition: partition)

        let response = try await api.send(
            "api/mobile/v1/giving/\(churchSlug)/funds",
            ifNoneMatch: cached?.etag,
            as: GivingHome.self
        )

        if response.notModified, let cached { return cached.value }
        guard let value = response.value else {
            throw APIError(code: .unavailable, message: L.givingUnavailableBody)
        }
        try? await cache.store(
            CacheEntry(value: value, etag: response.etag, storedAt: Date()),
            name: key,
            partition: partition
        )
        return value
    }

    // MARK: - Giving

    /// Starts, or resumes, one logical donation.
    ///
    /// The attempt id is the client's, is persisted before this is called, and is
    /// re-sent unchanged after any interruption — which is what makes the server
    /// return the intent it already created rather than a second one.
    public func startDonation(_ attempt: DonationAttempt) async throws -> DonationSession {
        let response = try await api.send(
            "api/mobile/v1/giving/donate",
            method: .post,
            body: StartDonationRequest(
                churchSlug: attempt.churchSlug,
                fundId: attempt.fundID,
                amountCents: attempt.amountCents,
                clientAttemptId: attempt.clientAttemptID
            ),
            as: DonationSession.self
        )
        return try require(response.value)
    }

    /// What the **server** believes happened. Never cached.
    public func status(
        churchSlug: String,
        attemptID: String
    ) async throws -> DonationStatusResult {
        let response = try await api.send(
            "api/mobile/v1/giving/\(churchSlug)/status/\(attemptID)",
            as: DonationStatusResult.self
        )
        return try require(response.value)
    }

    public func history(
        churchSlug: String,
        before: String? = nil
    ) async throws -> GivingHistoryPage {
        var params: [String: String] = [:]
        if let before { params["before"] = before }
        let response = try await api.send(
            "api/mobile/v1/giving/\(churchSlug)/history",
            query: params,
            as: GivingHistoryPage.self
        )
        return try require(response.value)
    }

    public func receipt(
        churchSlug: String,
        attemptID: String
    ) async throws -> GivingReceipt {
        let response = try await api.send(
            "api/mobile/v1/giving/\(churchSlug)/receipt/\(attemptID)",
            as: GivingReceipt.self
        )
        return try require(response.value)
    }

    /// A response with no body where one was required.
    ///
    /// These routes are all `no-store`, so a 304 is impossible and a missing
    /// value means the transport failed in a way the decoder could not see.
    /// Reported rather than papered over with a default.
    private func require<T>(_ value: T?) throws -> T {
        guard let value else {
            throw APIError(code: .unavailable, message: L.givingUnavailableBody)
        }
        return value
    }
}

File: apps/faithform-android/core/giving/src/main/kotlin/io/faithform/app/giving/GivingClient.kt
package io.faithform.app.giving

import io.faithform.app.contract.DonationSession
import io.faithform.app.contract.DonationStatusResult
import io.faithform.app.contract.GivingHistoryPage
import io.faithform.app.contract.GivingHome
import io.faithform.app.contract.GivingReceipt
import io.faithform.app.contract.MobileErrorCode
import io.faithform.app.contract.StartDonationRequest
import io.faithform.app.network.ApiClient
import io.faithform.app.network.ApiException
import io.faithform.app.network.FaithFormJson
import io.faithform.app.network.MobileSuccess
import io.faithform.app.network.ProjectionCache
import io.faithform.app.storage.CachePartition
import kotlinx.serialization.KSerializer

/**
 * Reads the giving projections and starts gifts.
 *
 * Mirrors `GivingClient.swift` route for route:
 *
 * | call              | route                                                | cached |
 * |-------------------|------------------------------------------------------|--------|
 * | [home]            | `GET  api/mobile/v1/giving/{slug}/funds`             | yes, ETag |
 * | [startDonation]   | `POST api/mobile/v1/giving/donate`                   | never |
 * | [status]          | `GET  api/mobile/v1/giving/{slug}/status/{attempt}`  | never |
 * | [history]         | `GET  api/mobile/v1/giving/{slug}/history?before`    | never |
 * | [receipt]         | `GET  api/mobile/v1/giving/{slug}/receipt/{attempt}` | never |
 *
 * ## What is cached, and what is not
 *
 * The fund list is a church's published configuration and changes rarely, so it
 * is cached under the caller's partition and revalidated with its ETag. **Nothing
 * else is.** A donation session carries a client secret, a status is the one
 * thing where a stale answer is actively harmful, and a person's giving history
 * is not something to leave in a cache on a shared device. Those routes answer
 * `no-store`, and this client does not second-guess them.
 *
 * ## Why `donate` carries no Idempotency-Key header
 *
 * The server makes a retry safe with `clientAttemptId` in the body instead: an
 * attempt id is persisted before the request and survives an app kill, a sheet
 * that never returned, and a person who comes back an hour later. A header
 * generated per request survives none of those.
 */
class GivingClient(
    private val api: ApiClient,
    private val cache: ProjectionCache,
) {
    suspend fun home(churchSlug: String, partition: CachePartition): GivingHome =
        cache.revalidate(
            api = api,
            path = "api/mobile/v1/giving/$churchSlug/funds",
            serializer = GivingHome.serializer(),
            name = "giving.funds.$churchSlug",
            partition = partition,
        )

    suspend fun cachedHome(churchSlug: String, partition: CachePartition) =
        cache.load("giving.funds.$churchSlug", partition, GivingHome.serializer())

    /** Starts, or resumes, one logical donation. */
    suspend fun startDonation(attempt: DonationAttempt): DonationSession = required(
        api.send(
            path = "api/mobile/v1/giving/donate",
            serializer = MobileSuccess.serializer(DonationSession.serializer()),
            method = "POST",
            body = FaithFormJson.encodeToString(
                StartDonationRequest.serializer(),
                StartDonationRequest(
                    churchSlug = attempt.churchSlug,
                    fundId = attempt.fundId,
                    amountCents = attempt.amountCents,
                    clientAttemptId = attempt.clientAttemptId,
                ),
            ),
        ).value,
    )

    /** What the **server** believes happened. Never cached. */
    suspend fun status(churchSlug: String, attemptId: String): DonationStatusResult =
        get("api/mobile/v1/giving/$churchSlug/status/$attemptId", DonationStatusResult.serializer())

    suspend fun history(churchSlug: String, before: String? = null): GivingHistoryPage = required(
        api.send(
            path = "api/mobile/v1/giving/$churchSlug/history",
            serializer = MobileSuccess.serializer(GivingHistoryPage.serializer()),
            query = before?.let { mapOf("before" to it) } ?: emptyMap(),
        ).value,
    )

    suspend fun receipt(churchSlug: String, attemptId: String): GivingReceipt =
        get("api/mobile/v1/giving/$churchSlug/receipt/$attemptId", GivingReceipt.serializer())

    private suspend fun <T> get(path: String, serializer: KSerializer<T>): T =
        required(api.send(path = path, serializer = MobileSuccess.serializer(serializer)).value)

    /**
     * A response with no body where one was required. These routes are all
     * `no-store`, so a 304 is impossible and a missing value means the transport
     * failed in a way the decoder could not see — reported, not defaulted.
     */
    private fun <T> required(value: T?): T = value ?: throw ApiException(
        code = MobileErrorCode.UNAVAILABLE,
        displayMessage = "Giving is unavailable right now.",
        retryable = true,
    )
}
